データベース設計

Yuichi-Abe-dev opened the pull request:
レビューをお願い致します。

## What
データベースの設計を示すREADMEとER図の作成

## Why
データベースの設計のため

ER図
https://gyazo.com/45be3abc9f45227ab0e592102047662b

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

Yuichi-Abe-dev committed:
Update README.md

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+| -------- | ---- | --------- |
+|nickname  |string|null: false|
+|email     |string|null: false, unique: true|
+|password  |string|null: false|
```
passwordのカラム名はdevise側がデフォルトで用意しているencrypted_passwordとしましょう。

理由
deviseを使用する場合、新規登録の際にpasswordというキーに対して入力された値は暗号化されてencrypted_passwordというカラムに保存されます。
そのため、このencrypted_passwordというカラムをpasswordカラムに変更してしまうとエラーが発生してしまいます。
DB設計の段階からdeviseの仕様に合わせておくことで、エラーを未然に防ぐことができます。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+|nickname  |string|null: false|
+|email     |string|null: false, unique: true|
+|password  |string|null: false|
+|first_name|string|null: false|
+|last_name |string|null: false|
```
お名前カナ(姓)、お名前カナ(名)を保存するカラムも用意しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
-* Database initialization
+|Column       |Type         |Options    |
+| --------    | ----        | --------- |
+|image        |ActiveStorage|null: false|
```
24行目
こちらのカラムですが、今回の実装では商品出品の画像投稿において
記述していただいている通りactive_storageを使用します。
READMEのテーブル設計においては
itemsテーブルからimageカラムは削除しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+|image        |ActiveStorage|null: false|
+|name         |string       |null: false|
+|description  |text         |null: false, unique: true|
+|category     |string       |null: false|
+|condition    |string       |null: false|
+|postage      |string       |null: false|
+|prefecture   |string       |null: false|
+|shipping_date|date         |null: false|
```
itemsテーブルのactive_hashを使用するカラムは
カラム名の末尾に_idを付けましょう。
例：category_id

またID番号を保存するカラムのため
integer型(整数)としておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+|postage      |string       |null: false|
+|prefecture   |string       |null: false|
+|shipping_date|date         |null: false|
+|price        |int          |null: false|
```
32行目
実際のマイグレーションファイルに記述するように
Typeは`integer`と記述しておくとより良いかと思います！

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+|prefecture   |string       |null: false|
+|shipping_date|date         |null: false|
+|price        |int          |null: false|
+|user_id      |references   |null: false, foreign_key: true|
```
外部キーをreferences型で記述する際は、カラム名についている_idは不要となりますので省略しましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+|shipping_date|date         |null: false|
+|price        |int          |null: false|
+|user_id      |references   |null: false, foreign_key: true|
+|purchase_id  |references   |foreign_key: true            |
```
３４行目
itemsテーブルにおいては
purchasesテーブルに対する外部キーカラムは不要となるので削除しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+## purchasesテーブル
+|Column      |Type         |Options    |
+| --------   | ----        | --------- |
+|user_id     |references   |null: false, foreign_key: true|
+|item_id     |references   |null: false, foreign_key: true|
```
こちらも外部キーをreferences型で記述する際は、カラム名についている_idは不要となりますので省略しましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+| --------   | ----        | --------- |
+|user_id     |references   |null: false, foreign_key: true|
+|item_id     |references   |null: false, foreign_key: true|
+|address_id  |references   |null: false, foreign_key: true|
```
purchasesテーブルにおいては
Addressesテーブルに対する外部キーカラムは不要となるので削除しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+|Column         |Type     |Options    |
+| --------      | ----    | --------- |
+|postal_code    |string   |null: false|
+|prefecture     |string   |null: false|
```
５８行目
こちらもactive_hashを使用するカラムとなるため
カラム名の末尾に_idを付けましょう。

また、Typeもintegerに修正しておきましょう。

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！
コメントいたしましたので、ご確認をお願いいたします。

Yuichi-Abe-dev commented:
> レビュー依頼ありがとうございます！
> コメントいたしましたので、ご確認をお願いいたします。

@CHURO-MAU 
レビューありがとうございます!🙇‍♂️
内容確認の上、修正致します。

Yuichi-Abe-dev committed:
Update README.md

パスワードのカラム名をencrypted_passwordに変更

Yuichi-Abe-dev committed:
Update README.md

お名前カナ(姓)、お名前カナ(名)を保存するカラムを追加

Yuichi-Abe-dev committed:
Update README.md

imageカラムを削除

Yuichi-Abe-dev committed:
Update README.md

active_hashを使用するカラムの名前とタイプの変更

Yuichi-Abe-dev committed:
Update README.md

active_hashのカラムのタイプをintegerに変更

Yuichi-Abe-dev committed:
Update README.md

intをintegerに変更

Yuichi-Abe-dev committed:
Update README.md

- itemsテーブルのuser_idをuser変更
- itemsテーブルのpurchasesテーブルへの参照を削除

Yuichi-Abe-dev committed:
Update README.md

purchasesテーブル
- 外部キーをreferences型で記述しているカラム名の_idを削除
- Addressesテーブルに対する外部キーカラムを削除

Yuichi-Abe-dev committed:
Update README.md

Addressesテーブル
- prefectureのカラム名をprefecture_id、Typeをintegerに変更
- phone_numberのTypeをintからintegerに変更

Yuichi-Abe-dev committed:
Update README.md

Addressesテーブルにpurchaseカラムを追加

Yuichi-Abe-dev commented:
@CHURO-MAU 
データベースを修正致しました。レビューをお願い致します。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,65 @@
+|Column          |Type         |Options    |
+| -------------- | ----        | --------- |
+|name            |string       |null: false|
+|description     |text         |null: false, unique: true|
```
下記修正後、マージしてきましょう！

商品説明に一意性制約は不要かと思われますので、
unique: trueは削除していただければと思います。

Yuichi-Abe-dev replied:
> レビュー依頼ありがとうございます！
> LGTMとさせていただきます。
> DB設計お疲れ様でした。
> 
> マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
> その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
> 引き続き実装頑張っていきましょう！

レビューありがとうございます！
修正の上、マージします。

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます。
DB設計お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう！

Yuichi-Abe-dev committed:
Update README.md

itemsテーブル のdescriptionのoptionのunique: true
